File: clients/go/transport.go
package dotprompt

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"
)

type Transport struct {
	baseURL    string
	apiKey     string
	timeout    time.Duration
	maxRetries int
	httpClient *http.Client
}

type TransportOptions struct {
	BaseURL    string
	APIKey     string
	Timeout    time.Duration
	MaxRetries int
}

func NewTransport(opts TransportOptions) *Transport {
	if opts.BaseURL == "" {
		opts.BaseURL = "http://localhost:4041"
	}
	if opts.Timeout == 0 {
		opts.Timeout = 30 * time.Second
	}
	if opts.MaxRetries == 0 {
		opts.MaxRetries = 3
	}

	return &Transport{
		baseURL:    strings.TrimRight(opts.BaseURL, "/"),
		apiKey:     opts.APIKey,
		timeout:    opts.Timeout,
		maxRetries: opts.MaxRetries,
		httpClient: &http.Client{Timeout: opts.Timeout},
	}
}

func (t *Transport) request(ctx context.Context, method, path string, body any) (map[string]any, error) {
	var reqBody io.Reader
	if body != nil {
		jsonBody, err := json.Marshal(body)
		if err != nil {
			return nil, fmt.Errorf("failed to marshal request body: %w", err)
		}
		reqBody = strings.NewReader(string(jsonBody))
	}

	url := t.baseURL + path
	var req *http.Request
	var err error

	if reqBody != nil {
		req, err = http.NewRequestWithContext(ctx, method, url, reqBody)
	} else {
		req, err = http.NewRequestWithContext(ctx, method, url, nil)
	}
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Content-Type", "application/json")
	if t.apiKey != "" {
		req.Header.Set("Authorization", "Bearer "+t.apiKey)
	}

	var lastErr error
	for attempt := 0; attempt <= t.maxRetries; attempt++ {
		if attempt > 0 {
			sleepDuration := time.Duration(1<<uint(attempt-1)) * 200 * time.Millisecond
			time.Sleep(sleepDuration)
		}

		resp, err := t.httpClient.Do(req)
		if err != nil {
			lastErr = err
			if ctx.Err() != nil {
				return nil, ctx.Err()
			}
			continue
		}
		defer resp.Body.Close()

		if resp.StatusCode >= 400 {
			return nil, t.handleError(resp)
		}

		var result map[string]any
		if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
			return nil, fmt.Errorf("failed to decode response: %w", err)
		}
		return result, nil
	}

	return nil, fmt.Errorf("request failed after %d attempts: %w", t.maxRetries+1, lastErr)
}

func (t *Transport) handleError(resp *http.Response) error {
	status := resp.StatusCode
	body, _ := io.ReadAll(resp.Body)
	message := string(body)

	var errType string
	var errMsg string
	if len(body) > 0 {
		var errData map[string]any
		if json.Unmarshal(body, &errData) == nil {
			if v, ok := errData["error"].(string); ok {
				errType = v
			}
			if v, ok := errData["message"].(string); ok {
				errMsg = v
			}
		}
	}

	if errMsg == "" {
		errMsg = message
	}
	if errMsg == "" {
		errMsg = "Unknown error"
	}

	switch status {
	case 404:
		return &PromptNotFoundError{PromptName: path, Message: errMsg}
	case 422:
		if strings.Contains(errMsg, "missing") || strings.Contains(errMsg, "required") {
			return &MissingRequiredParamsError{Message: errMsg}
		}
		return &ValidationError{Message: errMsg}
	case 500, 502, 503:
		return &ServerError{StatusCode: status, Message: errMsg}
	default:
		return &APIClientError{StatusCode: status, Message: errMsg}
	}
}

func (t *Transport) stream(ctx context.Context, path string) (*http.Response, error) {
	url := t.baseURL + path
	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
	if err != nil {
		return nil, err
	}

	if t.apiKey != "" {
		req.Header.Set("Authorization", "Bearer "+t.apiKey)
	}

	resp, err := t.httpClient.Do(req)
	if err != nil {
		return nil, err
	}

	if resp.StatusCode >= 400 {
		return nil, t.handleError(resp)
	}

	return resp, nil
}

File: clients/typescript/src/transport.ts
import { 
  ConnectionError, 
  TimeoutError, 
  ServerError, 
  APIClientError, 
  PromptNotFoundError,
  DotPromptError 
} from "./errors.js";
import { sleep } from "./utils.js";

/**
 * Options for a single HTTP request.
 */
export interface RequestOptions {
  method?: "GET" | "POST" | "PUT" | "DELETE";
  headers?: Record<string, string>;
  body?: any;
  timeout?: number;
}

/**
 * Transport layer for dot-prompt API requests.
 */
export class Transport {
  public baseUrl: string;
  public apiKey?: string;
  public timeout: number;
  public maxRetries: number;

  constructor(options: {
    baseUrl?: string;
    apiKey?: string;
    timeout?: number;
    maxRetries?: number;
  } = {}) {
    this.baseUrl = options.baseUrl || "http://localhost:4041";
    this.apiKey = options.apiKey;
    this.timeout = options.timeout || 30000; // default 30s
    this.maxRetries = options.maxRetries !== undefined ? options.maxRetries : 3;
  }

  /**
   * Helper to perform a request with retry logic.
   *
   * @param path - The API path (e.g. '/api/prompts').
   * @param options - Request options.
   * @returns T - The parsed JSON response.
   */
  public async request<T = any>(
    path: string, 
    options: RequestOptions = {}
  ): Promise<T> {
    let lastError: Error | null = null;

    for (let attempt = 0; attempt <= this.maxRetries; attempt++) {
      try {
        if (attempt > 0) {
          // Exponential backoff: 200ms, 400ms, 800ms...
          await sleep(Math.pow(2, attempt) * 100);
        }

        return await this._doRequest<T>(path, options);

      } catch (error: any) {
        lastError = error;

        // Don't retry client-side errors (4xx) or PromptNotFoundError
        if (error instanceof APIClientError) {
          throw error;
        }

        // Retry on 5xx or ConnectionError/TimeoutError
        if (attempt === this.maxRetries) {
          throw error;
        }
      }
    }

    throw lastError || new DotPromptError("Request failed after retries");
  }

  /**
   * Internal fetch wrapper.
   *
   * @param path - The API path.
   * @param options - Request options.
   * @returns T - The parsed JSON response.
   */
  private async _doRequest<T>(
    path: string, 
    options: RequestOptions = {}
  ): Promise<T> {
    const url = `${this.baseUrl}${path}`;
    const controller = new AbortController();
    const id = setTimeout(() => controller.abort(), options.timeout || this.timeout);

    const headers: Record<string, string> = {
      "Content-Type": "application/json",
      ...options.headers,
    };

    if (this.apiKey) {
      headers["Authorization"] = `Bearer ${this.apiKey}`;
    }

    try {
      const response = await fetch(url, {
        method: options.method || "GET",
        headers,
        body: options.body ? JSON.stringify(options.body) : undefined,
        signal: controller.signal,
      });

      clearTimeout(id);

      if (!response.ok) {
        return this._handleError(response);
      }

      return (await response.json()) as T;

    } catch (error: any) {
      clearTimeout(id);

      if (error.name === "AbortError") {
        throw new TimeoutError("Request timed out");
      }

      if (error.code === "ECONNREFUSED" || error.message.includes("fetch failed")) {
        throw new ConnectionError(`Could not connect to dot-prompt server at ${this.baseUrl}: ${error.message}`);
      }

      throw error;
    }
  }

  /**
   * Internal error mapper for HTTP responses.
   *
   * @param response - The Fetch API response object.
   */
  private async _handleError(response: Response): Promise<never> {
    const status = response.status;
    let message = "Unknown error";
    
    try {
      const body = await response.json();
      message = body.error || body.message || message;
    } catch {
      // Body not JSON, potentially raw text
      try {
        message = await response.text() || message;
      } catch {
        // Fallback to defaults
      }
    }

    if (status === 404) {
      throw new PromptNotFoundError(status, message);
    }

    if (status >= 500) {
      throw new ServerError(status, message);
    }

    throw new APIClientError(status, message);
  }
}
